### ISSUE TITLE
Http2_DataSentBeforeServerPreface_ProtocolError failed in CI

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#39966]
https://helix.dot.net/api/2019-06-17/jobs/12b2aa31-b7e0-4d12-b78d-35c3f4789b86/workitems/System.Net.Http.Functional.Tests/console
```   System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2.Http2_DataSentBeforeServerPreface_ProtocolError [FAIL]
      Assert.Contains() Failure
      Not found: PROTOCOL_ERROR
      In value:  System.Net.Http.HttpRequestException: Exception of type 'System.Net.Http.HttpRequestException' was thrown.
         at System.Net.Http.HttpConnectionBase.CreateRetryException() in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionBase.cs:line 56
         at System.Net.Http.Http2Connection.AddStream(Int32 streamId, HttpRequestMessage request) in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 1709
         at System.Net.Http.Http2Connection.SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken, Boolean mustFlush) in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 1200
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 1629
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 541
         at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs:line 33
         at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in /_/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 521
         at Xunit.Assert.RecordExceptionAsync(Func`1 testCode) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\Record.cs:line 88
      Stack Trace:
        /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(39,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.AssertProtocolErrorAsync[T](Task task, ProtocolErrors errorCode)
        /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(135,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.Http2_DataSentBeforeServerPreface_ProtocolError()
        --- End of stack trace from previous location where exception was thrown ---
```

**Comment by geoffkizer** [comment]
There is an inherent race between (a) sending a request and (b) connection shutdown or abort. We try to deal with this by checking _disposed (which indicates client-initiated shutdown) and _lastStreamId (which indicates server-initiated shutdown) in a couple places in the SendHeadersAsync path. If we find that the connection has started shutting down, we throw. 

In most cases, we throw a "retryable exception", which is an HttpRequestException created with the retry flag set. Typically this will be caught in HttpConnectionPool.SendWithRetryAsync, which will check the retry flag, and if set, will retry the request on a new HTTP2 connection. This handles the case of graceful shutdown, both client initiated (e.g. connection lifetime exceeded) and server initiated (i.e. GOAWAY). In particular, we throw this exception at these places:

https://github.com/dotnet/corefx/blob/175c4b7467f4da173c732f9aaf95882b3c426665/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1138
https://github.com/dotnet/corefx/blob/175c4b7467f4da173c732f9aaf95882b3c426665/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1709

The exception we throw in this case is not very useful; it has no message and it's not intended to be surfaced to the user. Usually, this retryable exception would cause us to retry the request on another connection. But we only go through the retry path when the connection is not newly created. We do this to avoid an infinite loop where the connection is failing immediately after being established (as here). So in this case, the "retryable exception" ends up getting surfaced to the user, even though we don't really expect it to.

There is another case in the SendHeadersAsync path that also handles connection shutdown -- the case where we fail because the _conncurrentStreams CreditManager has been disposed as part of connection shutdown. In this case, after we have determined that the connection is shutting down, we also check _abortException and, if it is set (meaning connection failure, not graceful shutdown), we throw a non-retryable exception for the connection abort. See here:

https://github.com/dotnet/corefx/blob/175c4b7467f4da173c732f9aaf95882b3c426665/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1116

I believe that in this test, we are usually hitting this path and throwing the abort exception, but occasionally, due to the inherent race here, we hit one of the other paths above and throw a retryable exception, which ends up getting surfaced to the user as seen above.

It's not clear that throwing a non-retryable exception in this case is completely correct either. Since we haven't started sending the request yet, there's no reason this exception can't be retryable as well. But the code as it stands precludes this.

So I think we need to do a couple things here. First, we should be consistent about the exception we throw in all three of the paths referenced above. Second, when the connection is aborting (i.e. _abortException is set), then in all these cases we should throw a retryable HttpRequestException with the _abortException as the inner exception (or possibly inner inner exception, to match other places where we end up throwing due to _abortException). This will allow the request to be retried unless it's a new connection, in which case we will surface the exception to the user. Third, for graceful shutdown scenarios, we need to throw a meaningful retryable exception for the unusual case where this exception could be surfaced because it's a new connection. For example, if the server completes the connection handshake and then immediately sends a GOAWAY -- we don't want to retry in this case, and so we need to surface a meaningful exception like "The server shut down the connection" that at least gives some information about what's happening.

**Comment by geoffkizer** [comment]
Reopen for 3.0

**Comment by karelz** [comment]
Fixed in 3.0 by PR #40028 and in master (5.0) in PR #39985
